crates/shim/src/asynchronous/publisher.rs

```diff
@@ -18,34 +18,110 @@ use std::os::unix::io::RawFd;
 
 use async_trait::async_trait;
 use containerd_shim_protos::{
-    api::Empty,
+    api::{Empty, Envelope},
     protobuf::MessageDyn,
     shim::events,
     shim_async::{Client, Events, EventsClient},
     ttrpc,
     ttrpc::{context::Context, r#async::TtrpcContext},
 };
+use log::debug;
+use tokio::sync::{mpsc, Mutex};
 
 use crate::{
-    error::Result,
+    asynchronous::Arc,
+    error::{self, Result},
     util::{asyncify, connect, convert_to_any, timestamp},
 };
+//Mas queue size
+const QUEUE_SIZE: i64 = 1024;
+//Max try five times
+const MAX_REQUEUE: i64 = 5;
 
 /// Async Remote publisher connects to containerd's TTRPC endpoint to publish events from shim.
 pub struct RemotePublisher {
-    client: EventsClient,
+    pub address: String,
+    client: Arc<Mutex<EventsClient>>,
+    sender: mpsc::Sender<Item>,
+    receiver: Arc<Mutex<mpsc::Receiver<Item>>>,
 }
 
+#[derive(Clone)]
+pub struct Item {
+    ev: Envelope,
+    ctx: Context,
+    count: i64,
+}
 impl RemotePublisher {
     /// Connect to containerd's TTRPC endpoint asynchronously.
     ///
     /// containerd uses `/run/containerd/containerd.sock.ttrpc` by default
     pub async fn new(address: impl AsRef<str>) -> Result<RemotePublisher> {
-        let client = Self::connect(address).await?;
+        let client = Self::connect(address.as_ref()).await?;
+        let (sender, receiver) = mpsc::channel::<Item>(QUEUE_SIZE as usize);
+        let rt = RemotePublisher {
+            address: address.as_ref().to_string(),
+            client: Arc::new(Mutex::new(EventsClient::new(client))),
+            sender,
+            receiver: Arc::new(Mutex::new(receiver)),
+        };
+        rt.process_queue().await;
```

Comment:
You can pass receiver directly to the process queue thread. This way you can remove Arc and Mutex.

```suggestion
        rt.process_queue(client).await;
```

Same applies to client. mpsc assumes one receiver, so you can use client there without locking.

Reply:
Use the Arc and Mutex is that the other thread will use it.

Reply:
Right, you don't need it in the other thread. Just the process queue thread.